Introduce collaborative documents types- update models and seeders

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/wp/66325
https://community.openproject.org/wp/66597

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Create models and basic data seeders for document collaboration workflows

# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->


* Add document types
* Existing document categories are migrated to new document types model

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
Add `CollaborativeDocument` as a Delegated `Document` type

`CollaborativeDocuments` are an extension of the "File Based" `Document` model.
And have the following additional properties:

  * Type
  * Status
  * Assignee
  * Responsible
  * Due Date

See [Rails Delegated Types](https://api.rubyonrails.org/classes/ActiveRecord/DelegatedType.html)

akabiru committed:
Define dedicated document type and status models

akabiru committed:
Relocate document category seeds to the documents module

akabiru committed:
Define document type, status and workflow seeders

akabiru committed:
Use attribute model key in place of the hardcoded one

akabiru committed:
Add dedicated `DocumentWorkflow` model

akabiru committed:
Replace document status colors with primer label variants

akabiru committed:
Normalize document types for data consistency

akabiru committed:
Move `CollaborativeDocument#type` to base `Document` model

Allows seeding types for legacy document types, and reuse on any
other documentable delegated types.

akabiru committed:
Move `CollaborativeDocument#author` to base `Document` model

Allow reuse in other documentable delegated types. Defined as optional
as historical records did not have an author.

akabiru committed:
Do not assume "Standard" as the default type

"Legacy" documents will remain unchanged.

akabiru committed:
Change Document category to nullable

Document categories (enumerations) are a legacy grouping mechanism. The new
document types will use `DocumentType` model.

akabiru committed:
Join collabarative documents into existing model

  * Migrate existing document categories to types
  * Update document type references to match previous categories
  * Add Collabboration properties to `documents` with 'legacy' `kind` for existing documents

An attempt for a smoother transition to "collaborative" documents

akabiru committed:
Define `Document` reference replacement on principal deletion

   - [x] Add proper relations with inverse dependent strategies
   - [x] Add behaviour tests to `Principal::DeleteJob`
   - [x] Replace references with deleted user

brunopagno commented on modules/documents/app/models/document.rb:
```diff
@@ -29,8 +29,19 @@
 #++
 
 class Document < ApplicationRecord
+  enum :kind, {
+    legacy: "legacy",
+    collaborative: "collaborative"
+  }
```
I might be missing some product context here, but is legacy what we were calling "document based"? Or are we misisng an entry here that will come in the future?

brunopagno replied:
So, as we don't have the file based one implemented, there's no need to add it

akabiru replied:
genau... yet 🙂

brunopagno commented on config/initializers/replace_references_service.rb:
```diff
@@ -37,6 +37,7 @@
       "CostEntry" => %i[logged_by_id user_id],
       "CostQuery" => :user_id,
       "::Doorkeeper::Application" => :owner_id,
+      "Document" => %i[author_id assigned_to_id responsible_id],
```
If you have the patience to draw how this service is used I'm curious.

_I might have a look on my own later, though_

akabiru replied:
Sure, this service replaces deleted user references with a "delete user" reference-  mitigates user foreign key constraint violation errors. https://github.com/opf/openproject/blob/03e187907db356c0173246a7cdfe8bf75d3183de/app/workers/principals/delete_job.rb#L48-L53

brunopagno commented on modules/documents/app/models/document_workflow.rb:
```diff
@@ -0,0 +1,36 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class DocumentWorkflow < ApplicationRecord
+  belongs_to :type, class_name: "DocumentType"
+  belongs_to :old_status, class_name: "DocumentStatus"
+  belongs_to :new_status, class_name: "DocumentStatus"
+  belongs_to :role
+end
```
I guess asking here for the sake of completeness, but we talked about using Workflow + Status + Type and jus creating document specialisation for those models. Here we created them isolated.

Anything specific that skewed the decision? This looks absolutely fine to me, asking out of curiosity, because it felt like we could benefit from the original models.

brunopagno replied:
Additional context: https://community.openproject.org/projects/communicator-stream/work_packages/66325/activity#comment-1435066

The original models were quite overloaded with WorkPackage context, so this cleans up the relations

brunopagno reviewed (commented):
I have left a few questions looking to understand the context a bit better, but overall looks great 👏 👏 👏 👏 

Maybe worth tagging devops for reviewing the migrations?

brunopagno reviewed (approved):
After pair reviewing live I declare this PR approved! 🚢 🚢 🚢 🚢

akabiru commented on app/models/user.rb:
```diff
@@ -98,6 +98,7 @@ class User < Principal
            inverse_of: :user,
            dependent: :destroy
 
+  has_many :documents, foreign_key: :author_id, dependent: :destroy, inverse_of: :author
```
From discussion with @brunopagno - the document is owned by the project, not exclusively by the author. When the author is deleted, the reference should be replaced with the "Deleted User" as in https://github.com/opf/openproject/pull/19999/files#diff-9e11cdc5737091941d4a9f2f454a194e846b9964bdc1c5e2f66974d528bc5a81

```suggestion
  has_many :documents, foreign_key: :author_id, dependent: :nullify, inverse_of: :author
```

judithroth commented on modules/documents/app/models/document_type.rb:
```diff
@@ -0,0 +1,43 @@
+  has_many :documents, foreign_key: :type_id,
+                       dependent: :nullify,
+                       inverse_of: :type
+  has_many :workflows, class_name: "DocumentWorkflow",
```
Do we need that yet? Because there are so many decision still in flow, I would not create those models unless we really need them now if I was the person implementing this.

akabiru replied:
Yes, these are necessary [as per the specified seeding](https://community.openproject.org/wp/66325). Currently, what is beyond the scope is a UI to manage workflows and types.

judithroth replied:
I'm just hesitant to add code that we don't need and use yet (but I may be mistaken if we need it right now :) ). Because requirements change so fast and then it might make more work to change something. For models, changing is easy, so we may keep them, with everything that is seeded to a database, it might be better to wait. But you're deeper into the topic than I am, so your judgment will be fine.

akabiru replied:
You make a valid point. Initially, this feature was added because we were certain we would require the workflows in the “show/edit” page of the documents. However, with your proposal, it would be added once we have these “show” pages. This makes sense. I’ll commit a revert so that we can restore it later. I’ll wait a bit until the final product specification has been finalized. 👍🏾

judithroth commented on modules/documents/app/models/document.rb:
```diff
@@ -29,8 +29,19 @@
 
 class Document < ApplicationRecord
+  enum :kind, {
+    legacy: "legacy",
```
Well... I know we are all struggling to find a good name here and haven't found it yet, but "legacy" can get confusing in the future, when, one fine day, we will introduce even better documents than we are introducing now. Or if we decide to keep the "old" kind of documents, then it's also confusing to read something about "legacy" in the code, since it is still used. For me personally, "legacy" implies that it should no longer be used. I don't know if there was a decision yet to completely throw out / migrate the documents as they have been?

akabiru replied:
I agree that the name “legacy” can be misleading. As far as I understand, the working agreement was that we wanted backward compatibility for the current (legacy) documents module before we could set plans to deprecate it. This is similar to what happened with the “Classic” and “Dynamic” meetings that we have today—with classic meetings eventually deprecated.

In the event that this code persists for a while, as it often does in software, perhaps “Classic” would be a more fitting future proof alternative. 😄

judithroth replied:
:+1: "classic" would work better for me than "legacy"

judithroth commented on modules/documents/app/seeders/common.yml:
```diff
@@ -35,3 +35,80 @@ modules_permissions:
+  - reference: :document_default_type_report
+    t_name: Report
+
+document_statuses:
```
Are these agreed on yet? :sweat_smile:

akabiru replied:
Yes they are (or should I say _were_) 🙂 Ditto: https://github.com/opf/openproject/pull/19999#discussion_r2316591093

judithroth replied:
Like I said in the other comment, it's up to your judgment if it's a good idea to already implement this :)

judithroth commented on modules/documents/app/seeders/common.yml:
```diff
@@ -35,3 +35,80 @@ modules_permissions:
+    t_name: Finalised
+    color_variant: done
+
+document_workflows:
```
And these? Do we need that at this stage?

akabiru replied:
Ditto: https://github.com/opf/openproject/pull/19999#discussion_r2316591093

judithroth replied:
Here as well :)

judithroth reviewed (commented):
Hagen encouraged me to also take a look. Good and neat code as always :+1: Since I assume you know more about documents (old and new) than me, this are just thoughts and questions that crossed my mind while reading  :)

akabiru committed:
Remove "collaborative" columns from documents

akabiru committed:
Remove "document workflows"

akabiru committed:
Remove "document status"

akabiru committed:
Update document types seeding with positioning

akabiru committed:
Prefer "classic" over "legacy" for existing document kinds

brunopagno reviewed (dismissed):
👏

akabiru committed:
Remove `Document#author` reference as it is not needed yet

akabiru committed:
Enhance document type migration plus seeding logic

Balance between seeding in new types and retaining (migrated) categories.
Only seed additional types when users have 3 or fewer migrated types.

judithroth commented on modules/documents/db/migrate/20250820131839_add_collaboration_to_documents.rb:
```diff
@@ -0,0 +1,57 @@
+  def set_existing_documents_to_classic_kind
+    say_with_time "setting existing documents to 'classic' kind" do
+      execute <<~SQL.squish
+        UPDATE documents SET kind = 'classic'
```
I'm not sure about this.
For customers, this will work well, since they haven't added any collaborative documents yet. For developers and community, this will set all to "classic" and so later on they will try to load CKEditor, regardless if they were Experimental (=collaborative) documents or not.
If it is too complex to set those to collaborative, we need to at least warn everyone that this will mess up formerly experimental/collaborative documents.

judithroth replied:
In regard to this, we should probably with this PR add a mechanism that when `OPENPROJECT_FEATURE_BLOCK_NOTE_EDITOR` is set all newly created "Experimental" documents get the correct type, shouldn't we? Otherwise this could also lead to documents getting the wrong type, which will probably cause confusion later on.

akabiru replied:
Really nice catches, Judith, thank you! 🌟 I agree with you, will add the changes! 👍🏾

akabiru committed:
Ensure we retain previous "Experimental" types as collaborative

The "Experimental" type was used to test blocknote features. While not released yet,
let's not break developer workflows

akabiru committed:
Replace document category with type app-wide

akabiru committed:
Set default document type when none is specified

judithroth commented on modules/documents/db/migrate/20250820131839_add_collaboration_to_documents.rb:
```diff
@@ -34,18 +34,33 @@ def change
+    say_with_time "setting existing documents to appropriate kinds" do
+      # Set documents with "Experimental" type to collaborative kind
+      # These were likely created with OPENPROJECT_FEATURE_BLOCK_NOTE_EDITOR enabled
+      execute <<~SQL.squish
```
Is it not possible to check for the feature decision here?
If that would be possible we could do something like
1. Assign "classic" to all documents
2. if the feature decision `OpenProject::FeatureDecisions.block_note_editor_active?` is used, set all with document_type is Experimental to collaborative? 

I mean, probably nothing will go wrong with your current approach, I'm just trying to be cautious and avoid problems 😅 but if you feel that this will not be an issue and want to move faster, you can keep it like it is.

akabiru replied:
Good call, since "Experimental" was a special type that we came up with, just for testing blocknote- I didn't see a huge need to be too specific- but you make a good point. The effort is negligible and we don't want to break things for that one user that had created the same type for other use cases. 👍🏾

judithroth reviewed (commented):
Nice and thorough work! Thanks for your perserverance with always updating this to the latest decisions 👍 

I found one thing that I would have done differently, but I leave it up to you if you think this should be changed or is good enough. Also, if you want to change it, I leave it to you if you want me to review it again or not (I gladly review to make sure everything works as expected, but if you're confident and just want to get it done, I don't have to)
From my point of view, this LGTM 🎇 

---

Actually, I wanted to approve this but somehow that was not accepted. Maybe because of the failing tests? However, there are still specs failing and I think they need to be adapted (have set the kind to collaborative for documents in test to make them render blocknote instead of CKEditor)

judithroth reviewed (approved):
Thanks for being patient with my cautionous approach!
LGTM! 🥳 
This is a huge step forward.
